### PR TITLE
Refactor Configuration

### LINKED ISSUE #650: Refactor configuration
Adapt all startup parameters to support multiple protocol networks.

Options changes:

- **ethereum** —> **network-providers**
    - format: list of `network_identifier`:`network_provider_url`
    - still support the old argument name, `ethereum`, mark it as deprecated and move towards removing it
- **network-subgraph-urls**
    - format: list of `network_identifier`:`network_subgraph_url`
- **epoch-subgraph-urls**
    - format: list of `network_identifier`:`epoch_subgraph_url`
- note: `network_identifier` can support both numeric CAIP2 ids OR their string aliases
- Validations:
    - lists provided to the three protocol network options must have same length
    - user must provide all three above values for each `network_identifier`
    - if no `network_identifier` provided, fall back to older behavior, can only use one network in that case.

### PR DESCRIPTION
**Pull request by tilacog**
Note:
> https://github.com/graphprotocol/indexer/pull/659 refactors some changes from this PR, most notably the input validation module.
> Reviewing https://github.com/graphprotocol/indexer/pull/659 instead of this PR for the latest refactorings is advised.

1. Refactors configuration to accept the following options as lists instead of single values:
    - `--network-provider` (new name for `ethereum`, which is now an alias)
    - `--epoch-subgraph`
    - `--network-subgraph-endpoint`
    - `--network-subgraph-deployment`

2. Introduces:
    - a validation function `validateNetworkOptions` to sanitize the new cardinality of the network options.
    - a `input-parsing.ts` module with parsers for the new network identification format.

3. To ensure compatibility, this PR restricts its scope to argument parsing and does not incorporate the network multiplexing changes expected from #652. which will resolve all the `FIXMES` introduced here.

New dependencies:
- `parsimmon`: for the new `input-parsing.ts` module.
- `lodash-countby`: used in the new `validateNetworkOptions` function.

Closes #650

### CODE REVIEW

**Review by hopeyen — COMMENTED** [review]
looks pretty good! how do you think untaggedUrl would be used?

File: packages/indexer-agent/src/commands/start.ts [hunk 1]
```diff
@@ -808,3 +817,71 @@ async function validateNetworkId(
...
+    usedOptions += ', --network-subgraph-endpoint'
+  }
+  if (networkSubgraphDeployments !== undefined) {
+    arraysToCheck.push(networkSubgraphDeployments)
+    usedOptions += ', --network-subgraph-deployment'
+  }
+  usedOptions += ']'
+
+  // Check for consistent length across network options
+  const commonSize = new Set(arraysToCheck.map(a => a.length)).size
```
[the remark this answers is not in the document]
Comment (tilacog) [reply packages/indexer-agent/src/commands/start.ts:852]: Thanks for the review!
I believe I did a somewhat similar refactor in #659.

**Comment by tilacog** [comment]
> looks pretty good! how do you think untaggedUrl would be used?

`untagged` inputs will be used if an indexer prefers to stick to a single network configuration. 

Indexers will likely use untagged and single network configurations for the upcoming release, as using multiple networks will block the next database migration.

**Comment by tilacog** [comment]
https://github.com/graphprotocol/indexer/pull/659 refactors some changes from this PR, most notably the input validation module.
    Reviewing https://github.com/graphprotocol/indexer/pull/659 instead of this PR for the latest refactorings is advised.

**Comment by tilacog** [comment]
Resolved by https://github.com/graphprotocol/indexer/pull/668

**Pull request closed without merging**